x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java

```diff
@@ -451,12 +460,25 @@ public void deleteUser(final DeleteUserRequest deleteUserRequest, final ActionLi
      */
     void verifyPassword(String username, final SecureString password, ActionListener<AuthenticationResult> listener) {
         getUserAndPassword(username, ActionListener.wrap((userAndPassword) -> {
-            if (userAndPassword == null || userAndPassword.passwordHash() == null) {
+            if (userAndPassword == null) {
+                logger.trace(
+                    "user [{}] does not exist in index [{}], cannot authenticate against the native realm",
+                    username,
+                    securityIndex.aliasName()
+                );
+                listener.onResponse(AuthenticationResult.notHandled());
+            } else if (userAndPassword.passwordHash() == null) {
+                logger.debug("user [{}] in index [{}] does not have a password, cannot authenticate", username, securityIndex.aliasName());
```

Comment:
I think this needs to be the same level as the rest messages in this method. Any reason this is the only debug?

Reply:
Because this is the only one where the outcome isn't obvious in other ways.

- User doesn't exist is a fairly normal case.
- Success is obvious
- Password failure has a message in the result

But, an empty password hash is rare, surprising and not logged in any other way.
We could change it to return a message in the AuthenticationResult instead, but I don't think it should be pushed down to a trace.

Reply:
Thanks, understood. I agree

Reply:
Actually, it's impossible for `passwordHash` to be null.

This is the only place where we construct a `UserAndPassword` object, and it's not possible to have a null password there.
https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java#L601